### PR TITLE
Convert SearchResult, InteractiveAuth, PushProcessor and Scheduler to TS

### PR DESCRIPTION
**Pull request by t3chguy**

### CODE REVIEW

**Comment by t3chguy** [comment]
Blocked until after RC

File: src/@types/PushRules.ts
Comment (turt2live) [inline src/@types/PushRules.ts:93]: custom conditions are still possible though?

File: src/interactive-auth.ts
Comment (turt2live) [inline src/interactive-auth.ts:585]: we should type the error, or at least leave comments about the cast. `// TODO: Types` is fine imo

File: src/pushprocessor.ts
Comment (turt2live) [inline src/pushprocessor.ts:48]: type already exists as `PushRuleSet`.

File: src/pushprocessor.ts
Comment (turt2live) [inline src/pushprocessor.ts:56]: type already exists

File: src/pushprocessor.ts
Comment (turt2live) [inline src/pushprocessor.ts:125]: please can we put static functions below the constructor :(

File: src/scheduler.ts
Comment (turt2live) [inline src/scheduler.ts:291]: one day we should just use the logging library directly.

**Review by turt2live — CHANGES_REQUESTED** [review]

**Comment by turt2live** [comment]
gah, pressed buttons too quickly: overall the first readthrough of the diff looks fine. It's large enough where there's probably going to be a second read on re-review and a small chance of a third.

File: src/@types/PushRules.ts
Comment (turt2live) [reply src/@types/PushRules.ts:93]: discussed oob: removing this type makes life easier down the road, and realistically we don't care about custom conditions at this layer. Would be fine with a comment acknowledging this.

**Comment by t3chguy** [comment]
@turt2live please merge after approval along with https://github.com/matrix-org/matrix-react-sdk/pull/6323 - I will be on holiday

**Review by turt2live — APPROVED** [review]

**Pull request merged**
